### ISSUE TITLE
Hibernate Reactive with Panache: automatic `@WithSessionOnDemand` improvement or clarification

### ISSUE THREAD

**Issue by FroMage** [quarkusio/quarkus#39008]
### Describe the bug

The docs at https://quarkus.io/guides/hibernate-reactive-panache#transactions say:

> In some cases, the session is opened automatically on demand. For example, if a Panache entity method is invoked in a Jakarta REST resource method in an application that includes the `quarkus-resteasy-reactive` extension

This does mention that the REST resource method has to use a "Panache entity method", but it's not entirely clear to me. I interpreted it as "if you're using Panache entity methods within the context of REST resource methods" no matter how deep inside the method. But it's actually only if you use entity methods _directly_ in the REST resource class, and not via indirection of other classes. Also, it's also for repository users, but again, direct ones.

The exception that happens says:

```
java.lang.IllegalStateException: No current Mutiny.Session found
	- no reactive session was found in the Vert.x context and the context was not marked to open a new session lazily
	- a session is opened automatically for JAX-RS resource methods annotated with an HTTP method (@GET, @POST, etc.); inherited annotations are not taken into account
	- you may need to annotate the business method with @WithSession or @WithTransaction
```

Again, this does not mention that it's only automatic for direct users of Panache entity/repos, which is misleading.

Last but not least, this does not mention `@WithSessionOnDemand`, which is what we're talking about in the first place. It should.

We should improve the docs by listing exactly the circumstances in which `@WithSessionOnDemand` is added:

```
                    // Add @WithSessionOnDemand to a method that
                    // - is not static
                    // - is not synthetic
                    // - returns Uni
                    // - is declared in a class that uses a panache entity/repository
                    // - is annotated with @GET, @POST, @PUT, @DELETE ,@PATCH ,@HEAD or @OPTIONS
                    // - is not annotated with @ReactiveTransactional, @WithSession, @WithSessionOnDemand, or @WithTransaction
```

This is both for the guide, but also the annotation javadoc, because it's relevant there too.

Finally, we may want to improve this by making a graph of indirect users of Panache entities/repos, but probably this would include too many false positives and end up including everything.

**Comment by FroMage** [comment]
We could improve this by also adding `@WithSessionOnDemand` to rest endpoint methods whose signature contains Panache entities, as is the case in most endpoints that do not invoke entity operations, delegate them to a service, but obtain or return those entities via their signatures.

**Comment by mkouba** [comment]
> We could improve this by also adding `@WithSessionOnDemand` to rest endpoint methods whose signature contains Panache entities, as is the case in most endpoints that do not invoke entity operations, delegate them to a service, but obtain or return those entities via their signatures.

I believe that this use case is already covered because we add `@WithSessionOnDemand`  to all resource methods declared on a "known user" of an entity/repository. Where the "known user" is identified using the `IndexView#getKnownUsers()` method (which explicitly mentions _"the signature of any of the class's methods"_).

**Comment by FroMage** [comment]
Oh, yes, that makes sense indeed.

**Comment by mkouba** [comment]
> By the way, what's the difference between `@WithSession` a `@WithSessionOnDemand`? The javadocs are basically the same.

It's not the same 😄 . The javadoc of `@WithSessionOnDemand` mentions that a new session is only created **"when needed"**, i.e. is created lazily. For `@WithSession` a new session is created if it does not exist.

**Comment by DavideD** [comment]
Ok, that's why I said _basically_. 

But mainly I'm wondering if there's a use case where one would prefer to use `@WithSession` over `@WithSessionOnDemand`? Wouldn't it make more sense to have only one annotation (`@WithSession` that behaves like `@WithSessionOnDemand`)?

**Comment by FroMage** [comment]
The question is still valid, though. Perhaps we only need `@WithSessionOnDemand` (possibly renamed to `@WithSession`). What's the use-case for having a session opened even if it's never used?

**Comment by mkouba** [comment]
I don't think we should make it lazy by default. For users, it's more consistent if `@WithSession` maps to `Mutiny.SessionFactory#withSession()`. `@WithSessionOnDemand` is just an optimization for fwk.

**Comment by DavideD** [comment]
I would expect a user to only care that the session is available and managed when a method or the class are annotated. 
I don't think it matters if it's a wrapper around `SessionFactory#withSession` or some Quarkus magic. Unless maybe there are some performance considerations.

It also means having two annotations to test instead of one.

In any case, shouldn't the javadoc mention at least some of this?

**Comment by mkouba** [comment]
> In any case, shouldn't the javadoc mention at least some of this?

Mention what exactly? If you look [at our docs](https://quarkus.io/guides/hibernate-reactive-panache#transactions) we intentionally skip the `@WithSessionOnDemand` completely because in many cases a user does not have to do anything; otherwise `@WithSession` and `@WithTransaction` should do the trick.

But feel free to update the docs/javadoc if you believe it's important ;-).

**Comment by DavideD** [comment]
Even if you don't mention it in the doc, the annotation appears with the auto-completion in the IDE. And from the javadoc, it's unclear why one should use `@withSession` over `@WithSessionOnDemand`. When I saw it, I thought something was wrong or missing.

But, ok, this is by design, so fine by me.
